tests/lib/rules/hook-use-state.js

```diff
@@ -0,0 +1,233 @@
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+    },
+    {
+      code: `import { useState } from 'react';
+      const result = useState()`,
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+    },
+    {
+      code: `import { useState } from 'react';
+      const [, , extra1] = useState()`,
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+    },
+    {
+      code: `import { useState } from 'react';
+      const [, setColor] = useState()`,
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+    },
+    {
+      code: `import { useState } from 'react';
+      const { color } = useState()`,
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+    },
+    {
+      code: `import { useState } from 'react';
+      const [] = useState()`,
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+    },
+    {
+      code: `import { useState } from 'react';
+      const [, , , ,] = useState()`,
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+    },
+    {
+      code: `import { useState } from 'react';
+      const [color] = useState()`,
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+      output: `import { useState } from 'react';
+      const [color, setColor] = useState()`,
+    },
+    {
+      code: `import { useState } from 'react';
+      const [color, , extra1] = useState()`,
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+      output: `import { useState } from 'react';
+      const [color, setColor] = useState()`,
+    },
+    {
+      code: `import { useState } from 'react';
+      const [color, setColor, extra1, extra2, extra3] = useState()`,
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+      output: `import { useState } from 'react';
+      const [color, setColor] = useState()`,
+    },
+    {
+      code: `import { useState } from 'react';
+      const [, makeColor] = useState()`,
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+    },
+    {
+      code: `import { useState } from 'react';
+      const [color, setFlavor, extraneous] = useState()`,
+      errors: [{
+        message: 'useState call is not destructured into value + setter pair',
+      }],
+      output: `import { useState } from 'react';
+      const [color, setColor] = useState()`,
+    },
+    {
+      code: `import { useState } from 'react';
+      const [color, setFlavor] = useState()`,
```

Comment:
`useState` can only be called inside an SFC that React renders (or inside another hook), so i don't think these tests are valid unless they're inside a component or inside a custom hook

Reply:
Do you feel that these cases as-written should be flagged as invalid by this rule?
Or should I just add the dressing to these test cases?

Reply:
As written, the rule should ignore them, because they're not actually a valid hook call - it will throw at runtime already.

Also, eslint-plugin-react-hooks will, i believe, warn on it as-is.

Reply:
i thought we decided this should be ignored? or did we decide instead that its placement should be ignored.

Reply:
`eslint-plugin-react-hooks#rules-of-hooks` is almost entirely dedicated to detecting
* [where a hook is being called](https://github.com/facebook/react/blob/3dc41d8a2590768a6ac906cd1f4c11ca00417eee/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L91-L105)
* [whether it's called in an appropriate context](https://github.com/facebook/react/blob/3dc41d8a2590768a6ac906cd1f4c11ca00417eee/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L351-L354)

I see a few options here.
* *Do nothing*; flag this case as invalid, and rely on `eslint-plugin-react-hooks` to flag the bad usage
* *Do something simple*, like ignore top-level invocations of the hook
* *Do something ambitious* like port (most of) FBs RulesOfHooks over to this repo
* *Do something ambitious* like integrate FBs RulesOfHooks into `util/Components`
* *Do something ambitious* like extract FBs RulesOfHooks into a new module that can be shared by both repos
* *Do something **even more ambitious*** which I haven't thought of.

I'm in favor of *Do nothing*
The price of a mis-flag here is low; neither this rule nor `eslint-plugin-react-hooks#rules-of-hooks` attempt to **fix** this situation (suggestions-only), so it doesn't add any extra fixer passes.

Reply:
I checked out copying over pieces from rules-of-hooks, but its architected a bit differently.

Whereas this rule relies solely on visiting `CallExpression` nodes, `rules-of-hooks`:
 * [accumulates info about `CallExpressions`](https://github.com/facebook/react/blob/3dc41d8a2590768a6ac906cd1f4c11ca00417eee/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L516-L528)
 * and then does its heavy-lifting, including checking the `isInsideComponentOrHook` status [when the rule's `onCodePathEnd` callback fires](https://github.com/facebook/react/blob/3dc41d8a2590768a6ac906cd1f4c11ca00417eee/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L127-L142).